compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -449,6 +449,20 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                     }
                 }
             }
+            ExprKind::AddressOf { arg, .. } => {
+                if let ExprKind::Scope { value: arg, .. } = self.thir[arg].kind
+                // THIR desugars UNSAFE_STATIC into *&raw mut UNSAFE_STATIC, so take two steps
+                    && let ExprKind::Deref { arg } = self.thir[arg].kind
+                    && let ExprKind::StaticRef { .. } | ExprKind::ThreadLocalRef { .. } = self.thir[arg].kind
```

Comment:
Are you sure we want to include `ThreadLocalRef` here? Those actually have runtime behavior.

But I guess it's never unsafe to get the address of a thread-local so... this is probably scary but fine?

Reply:
Also what about extern statics? Is safe code allowed to cause a linker error?

Reply:
> Is safe code allowed to cause a linker error?

Yes. Linker errors aren't UB, they are just "very post-monomorphization".

Reply:
I don't like excluding them because I cannot come up with a good reason to do so aside from "spooky!" It feels inappropriately superstitious.

I however do not have a deep enough understanding of thread-local vagaries to be able to say confidently that there is no platform that would make this unsafe, so I am going to exclude them for now.